File: source_files/haskell/Q1-10-lists/p09.hs
-- Pack consecutive duplicates of list elements into sublists. 
-- If a list contains repeated elements they should be placed in seperate sublists. 

-- Using Data.List

import Data.List

pack :: Eq a => [a] -> [[a]]
pack = group

-- With list recursion. 
pack' :: Eq a => [a] -> [[a]]
pack' [] = [] 
pack' (x:[]) = [[x]]
pack' (x:xs) = (\((b:bs):ys) -> if x == b
                                       then (x:b:bs):ys
                                   else [x]:(b:bs):ys) (pack' xs)

-- With foldr.
pack'' :: Eq a => [a] -> [[a]]
pack'' list = foldr (\a ((b:bs):xs) -> if a == b
                                            then (a:b:bs):xs
                                       else [a]:(b:bs):xs) [[last list]] (init list)

File: ben/prolog/P01-P28-lists/p09.pl
% Pack consecutive duplicates of list elements into sublists.

pack_first([], _, [], []). 
pack_first([X|Xs], X, R, [X|Ys]) :-
    pack_first(Xs, X, R, Ys).
pack_first([X|Xs], Y, [X|Xs], []) :- 
    X \= Y.

pack([], []).
pack([Y|Ys], [X|Xs]) :-
    pack_first([Y|Ys], Y, L, X),
    pack(L, Xs).
